File: kant_dao/src/kant_dao/ros2_dao/ros2_pddl_proposition_dao.cpp

#include "kant_interfaces/msg/update_knowledge.hpp"

#include "kant_dao/ros2_dao/ros2_pddl_proposition_dao.hpp"

using namespace kant::dao::ros2_dao;

Ros2PddlPropositionDao::Ros2PddlPropositionDao(simple_node::Node *node) {

  this->dto_msg_parser =
      std::make_unique<kant::knowledge_base::parser::DtoMsgParser>();
  this->msg_dto_parser =
      std::make_unique<kant::knowledge_base::parser::MsgDtoParser>();

  // srv clients
  this->get_client =
      node->create_client<kant_interfaces::srv::GetPddlProposition>(
          "get_propositions");

  this->update_client =
      node->create_client<kant_interfaces::srv::UpdatePddlProposition>(
          "update_proposition");

  this->delete_all_client =
      node->create_client<std_srvs::srv::Empty>("delete_all_propositions");
}

std::vector<std::shared_ptr<kant::dto::PddlPropositionDto>>
Ros2PddlPropositionDao::ros2_get(int get_type, std::string predicate_name) {

  auto request =
      std::make_shared<kant_interfaces::srv::GetPddlProposition::Request>();

  request->predicate_name = predicate_name;
  request->get_type = get_type;

  this->get_client->wait_for_service();
  auto future = this->get_client->async_send_request(request);
  future.wait();
  auto result = future.get();

  std::vector<std::shared_ptr<kant::dto::PddlPropositionDto>>
      pddl_proposition_dto_list;

  for (const auto &pddl_proposition_msg : result->pddl_propositions) {
    auto pddl_proposition_dto =
        this->msg_dto_parser->proposition_msg_to_dto(pddl_proposition_msg);
    pddl_proposition_dto_list.push_back(pddl_proposition_dto);
  }

  return pddl_proposition_dto_list;
}

bool Ros2PddlPropositionDao::ros2_update(
    std::shared_ptr<kant::dto::PddlPropositionDto> pddl_proposition_dto,
    int update_type) {

  auto request =
      std::make_shared<kant_interfaces::srv::UpdatePddlProposition::Request>();

  request->pddl_proposition =
      this->dto_msg_parser->proposition_dto_to_msg(pddl_proposition_dto);
  request->update_konwledge.update_type = update_type;

  this->update_client->wait_for_service();
  auto future = this->update_client->async_send_request(request);
  future.wait();
  auto result = future.get();

  return result->success;
}

void Ros2PddlPropositionDao::ros2_delete_all() {

  auto request = std::make_shared<std_srvs::srv::Empty::Request>();
  this->delete_all_client->wait_for_service();
  auto future = this->delete_all_client->async_send_request(request);
  future.wait();
}

bool Ros2PddlPropositionDao::ros2_exists(
    std::shared_ptr<kant::dto::PddlPropositionDto> pddl_proposition_dto) {

  std::string predicate_name =
      pddl_proposition_dto->get_predicate()->get_name();

  for (const auto &dto : this->get_by_predicate(predicate_name)) {
    if (dto->equals(pddl_proposition_dto)) {
      return true;
    }
  }

  return false;
}

std::vector<std::shared_ptr<kant::dto::PddlPropositionDto>>
Ros2PddlPropositionDao::get_by_predicate(std::string predicate_name) {

  auto pddl_proposition_dto_list = this->ros2_get(
      kant_interfaces::srv::GetPddlProposition::Request::BY_PREDICATE,
      predicate_name);

  return pddl_proposition_dto_list;
}

std::vector<std::shared_ptr<kant::dto::PddlPropositionDto>>
Ros2PddlPropositionDao::get_goals() {

  auto pddl_proposition_dto_list = this->ros2_get(
      kant_interfaces::srv::GetPddlProposition::Request::GOALS, "");

  return pddl_proposition_dto_list;
}

std::vector<std::shared_ptr<kant::dto::PddlPropositionDto>>
Ros2PddlPropositionDao::get_no_goals() {

  auto pddl_proposition_dto_list = this->ros2_get(
      kant_interfaces::srv::GetPddlProposition::Request::NO_GOALS, "");

  return pddl_proposition_dto_list;
}

std::vector<std::shared_ptr<kant::dto::PddlPropositionDto>>
Ros2PddlPropositionDao::get_all() {

  auto pddl_proposition_dto_list = this->ros2_get(
      kant_interfaces::srv::GetPddlProposition::Request::ALL, "");

  return pddl_proposition_dto_list;
}

bool Ros2PddlPropositionDao::save(
    std::shared_ptr<kant::dto::PddlPropositionDto> pddl_proposition_dto) {

  if (!this->ros2_exists(pddl_proposition_dto)) {
    bool succ = this->ros2_update(pddl_proposition_dto,
                                  kant_interfaces::msg::UpdateKnowledge::SAVE);
    return succ;
  }

  return false;
}

bool Ros2PddlPropositionDao::update(
    std::shared_ptr<kant::dto::PddlPropositionDto> pddl_proposition_dto) {

  if (this->ros2_exists(pddl_proposition_dto)) {
    bool succ = this->ros2_update(pddl_proposition_dto,
                                  kant_interfaces::msg::UpdateKnowledge::SAVE);
    return succ;
  }

  return false;
}

bool Ros2PddlPropositionDao::save_update(
    std::shared_ptr<kant::dto::PddlPropositionDto> pddl_proposition_dto) {

  bool succ;
  if (!this->ros2_exists(pddl_proposition_dto)) {
    succ = this->save(pddl_proposition_dto);
  } else {
    succ = this->update(pddl_proposition_dto);
  }

  return succ;
}

bool Ros2PddlPropositionDao::delete_one(
    std::shared_ptr<kant::dto::PddlPropositionDto> pddl_proposition_dto) {
  bool succ = this->ros2_update(pddl_proposition_dto,
                                kant_interfaces::msg::UpdateKnowledge::DELETE);
  return succ;
}

bool Ros2PddlPropositionDao::delete_all() {

  this->ros2_delete_all();
  return true;
}

File: kant_dao/kant_dao/ros2_dao/ros2_pddl_proposition_dao.py

""" Ros2 Pddl Proposition Dao """

from typing import List
from simple_node import Node

from kant_dao.dao_interface import PddlPropositionDao
from kant_dto import PddlPropositionDto

from kant_msgs.srv import (
    UpdatePddlProposition,
    GetPddlProposition
)
from kant_msgs.msg import UpdateKnowledge
from std_srvs.srv import Empty

from kant_knowledge_base.parser import (
    DtoMsgParser,
    MsgDtoParser
)


class Ros2PddlPropositionDao(PddlPropositionDao):
    """ Ros2 Pddl Proposition Dao Class """

    def __init__(self, node: Node):

        PddlPropositionDao.__init__(self)

        self.node = node

        # parsers
        self.dto_msg_parser = DtoMsgParser()
        self.msg_dto_parser = MsgDtoParser()

        # srv clients
        self._get_client = self.node.create_client(
            GetPddlProposition, "get_propositions")

        self._update_client = self.node.create_client(
            UpdatePddlProposition, "update_proposition")

        self._delete_all_client = self.node.create_client(
            Empty, "delete_all_propositions")

    def _ros2_get(self,
                  get_type: int,
                  predicate_name: str = "") -> List[PddlPropositionDto]:
        """ ros2_get method

        Args:
            predicate_name (str): predicate name

        Returns:
            List[PddlPropositionDto]: list of PddlPropositionDto
        """

        req = GetPddlProposition.Request()

        req.predicate_name = predicate_name
        req.get_type = get_type

        self._get_client.wait_for_service()
        result = self._get_client.call(req)

        pddl_proposition_dto_list = []

        for pddl_proposition_msg in result.pddl_propositions:
            pddl_proposition_dto = self.msg_dto_parser.proposition_msg_to_dto(
                pddl_proposition_msg)
            pddl_proposition_dto_list.append(pddl_proposition_dto)

        return pddl_proposition_dto_list

    def _ros2_update(self,
                     pddl_proposition_dto: PddlPropositionDto,
                     update_type: int) -> bool:
        """ ros2_delete_all method

        Args:
            pddl_proposition_dto (PddlPropositionDto): PddlPropositionDto to update

        Returns:
            bool: succeed
        """

        req = UpdatePddlProposition.Request()

        req.pddl_proposition = self.dto_msg_parser.proposition_dto_to_msg(
            pddl_proposition_dto)
        req.update_konwledge.update_type = update_type

        self._update_client.wait_for_service()
        result = self._update_client.call(req)

        return result.success

    def _ros2_delete_all(self):
        """  delete_all method

        Returns:
            bool: succeed
        """

        req = Empty.Request()

        self._delete_all_client.wait_for_service()
        self._delete_all_client.call(req)

    def get_by_predicate(self, predicate_name: str) -> List[PddlPropositionDto]:
        """ get a PddlPropositionDto list with a given predicate name

        Args:
            predicate_name (str): pddl predicate name

        Returns:
            PddlPropositionDto: list of PddlPropositionDto of the pddl predicate name
        """

        pddl_proposition_dto_list = self._ros2_get(GetPddlProposition.Request.BY_PREDICATE,
                                                   predicate_name=predicate_name)

        return pddl_proposition_dto_list

    def get_goals(self) -> List[PddlPropositionDto]:
        """ get a PddlPropositionDto list of goals

        Returns:
            PddlPropositionDto: list of PddlPropositionDto goals
        """

        pddl_proposition_dto_list = self._ros2_get(
            GetPddlProposition.Request.GOALS)

        return pddl_proposition_dto_list

    def get_no_goals(self) -> List[PddlPropositionDto]:
        """ get a PddlPropositionDto list of no goals

        Returns:
            PddlPropositionDto: list of PddlPropositionDto no goals
        """

        pddl_proposition_dto_list = self._ros2_get(
            GetPddlProposition.Request.NO_GOALS)

        return pddl_proposition_dto_list

    def get_all(self) -> List[PddlPropositionDto]:
        """ get all PddlPropositionDto

        Returns:
            List[PddlPropositionDto]: list of all PddlPropositionDto
        """
        pddl_proposition_dto_list = self._ros2_get(
            GetPddlProposition.Request.ALL)

        return pddl_proposition_dto_list

    def _save(self, pddl_proposition_dto: PddlPropositionDto) -> bool:
        """ save a PddlPropositionDto
            if the PddlPropositionDto is already saved return False, else return True

        Args:
            pddl_proposition_dto (PddlPropositionDto): PddlPropositionDto to save

        Returns:
            bool: succeed
        """

        predicate_name = pddl_proposition_dto.get_predicate().get_name()
        if not pddl_proposition_dto in self.get_by_predicate(predicate_name):
            succ = self._ros2_update(
                pddl_proposition_dto, UpdateKnowledge.SAVE)
            return succ

        return False

    def _update(self, pddl_proposition_dto: PddlPropositionDto) -> bool:
        """ update a PddlPropositionDto
            if the PddlPropositionDto is not saved return False, else return True

        Args:
            pddl_proposition_dto (PddlPropositionDto): PddlPropositionDto to update

        Returns:
            bool: succeed
        """

        predicate_name = pddl_proposition_dto.get_predicate().get_name()
        if pddl_proposition_dto in self.get_by_predicate(predicate_name):
            succ = self._ros2_update(
                pddl_proposition_dto, UpdateKnowledge.SAVE)
            return succ

        return False

    def save(self, pddl_proposition_dto: PddlPropositionDto) -> bool:
        """ save or update a PddlPropositionDto
            if the PddlPropositionDto is not saved it will be saved, else it will be updated

        Args:
            pddl_proposition_dto (PddlPropositionDto): PddlPropositionDto to save or update

        Returns:
            bool: succeed
        """

        predicate_name = pddl_proposition_dto.get_predicate().get_name()
        if not pddl_proposition_dto in self.get_by_predicate(predicate_name):
            succ = self._save(pddl_proposition_dto)
        else:
            succ = self._update(pddl_proposition_dto)

        return succ

    def delete(self, pddl_proposition_dto: PddlPropositionDto) -> bool:
        """ delete a PddlPropositionDto
            if the PddlPropositionDto is not saved return False, else return True

        Args:
            pddl_proposition_dto (PddlPropositionDto): PddlPropositionDto to delete

        Returns:
            bool: succeed
        """

        succ = self._ros2_update(
            pddl_proposition_dto, UpdateKnowledge.DELETE)
        return succ

    def delete_all(self) -> bool:
        """ delete all pddl propositions

        Returns:
            bool: succeed
        """

        self._ros2_delete_all()

        return True
